site/public/js/notebook_builder/form-options-widget.js

```diff
@@ -0,0 +1,72 @@
+class FormOptionsWidget extends Widget {
+    constructor() {
+        super();
+    }
+
+    render() {
+        const heading_container = this.getHeadingContainer('Form Options');
+
+        const interactive_container = this.getInteractiveContainer();
+        interactive_container.innerHTML = this.getFormOptionsTemplate();
+
+        const container = document.createElement('div');
+        container.classList.add('notebook-builder-widget');
+        container.classList.add('form-options-widget');
+        container.appendChild(heading_container);
+        container.appendChild(interactive_container);
+
+        const save_button = container.querySelector('.save-button');
+        save_button.addEventListener('click', event => {
+            this.saveButtonAction();
+        })
+
+        return container;
+    }
+
+    /**
+     * Get the template used in the form options widget
+     *
+     * @returns {string}
+     */
+    getFormOptionsTemplate() {
+        return `
+        <div class="buttons">
+            <input type="button" class="save-button" value="Save">
+        </div>
+        <div class="status"></div>`;
+    }
+
+    /**
+     * Pack up all the configuration data from the GUI into a nice config.json and submit it to the server
+     */
+    saveButtonAction() {
+        const file = new File([JSON.stringify(notebook_builder.getJSON(), null, 2)], 'config.json', {type: "text/plain"});
+        const g_id = window.location.pathname.split('/').pop();
+        const url = buildCourseUrl(['notebook_builder', 'save']);
+
+        const form_data = new FormData();
+        form_data.append('config_upload', file, 'config.json');
+        form_data.append('csrf_token', csrfToken);
+        form_data.append('g_id', g_id);
+
+        const makeRequest = async () => {
+            const status_div = document.querySelector('.form-options-widget .status');
+            status_div.innerHTML = 'Saving...';
+
+            const response = await fetch(url, {method: 'POST', body: form_data});
```

Comment:
Instead of wrapping the `await` key `fetch` is async by default like ajax so you can use promises to handle this type of logic. I think it'll look something like this
```js
fetch(url, {
  method: 'POST',
  body: form_data,
})
.then(
   response => {
        JSON.parse(response)
   }
)
.then(data => {
  if(data.status === 'success){ 
     //foo
  }else{

  }
})
.catch((error) => {
  console.error('Error:', error);
});
```

Reply:
`await`/`async` is syntactic sugar over promises.  Any `async` function will always return a promise.  I think using `await`/`async` makes the code much more readable than using all then chaining `.then()`.

Reply:
Ok sounds good, I didn't realize `async` will always return a promise in that case I agree its more readable, maybe we can add this to the [coding style guide for JS](https://submitty.org/developer/coding_style_guide/javascript) @MasterOdin so we're consitent about using this.

personally I think we should start moving towards using `fetch` wherever we would use ajax in the past so having a consistent way to call it with promises would be nice.